pkg/query/selection.go

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package query
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/Peripli/service-manager/pkg/util/slice"
+
+	"github.com/Peripli/service-manager/pkg/web"
+)
+
+// Operator is a query operator
+type Operator string
+
+const (
+	// EqualsOperator takes two operands and tests if they are equal
+	EqualsOperator Operator = "="
+	// NotEqualsOperator takes two operands and tests if they are not equal
+	NotEqualsOperator Operator = "!="
+	// GreaterThanOperator takes two operands and tests if the left is greater than the right
+	GreaterThanOperator Operator = "gt"
+	// LessThanOperator takes two operands and tests if the left is lesser than the right
+	LessThanOperator Operator = "lt"
+	// InOperator takes two operands and tests if the left is contained in the right
+	InOperator Operator = "in"
+	// NotInOperator takes two operands and tests if the left is not contained in the right
+	NotInOperator Operator = "notin"
+	// EqualsOrNilOperator takes two operands and tests if the left is equal to the right, or if the left is nil
+	EqualsOrNilOperator Operator = "eqornil"
+)
+
+// IsMultiVariate returns true if the operator requires right operand with multiple values
+func (op Operator) IsMultiVariate() bool {
+	return op == InOperator || op == NotInOperator
+}
+
+// IsNullable returns true if the operator can check if the left operand is nil
+func (op Operator) IsNullable() bool {
+	return op == EqualsOrNilOperator
+}
+
+// IsNumeric returns true if the operator works only with numeric operands
+func (op Operator) IsNumeric() bool {
+	return op == LessThanOperator || op == GreaterThanOperator
+}
+
+var operators = []Operator{EqualsOperator, NotEqualsOperator, InOperator,
+	NotInOperator, GreaterThanOperator, LessThanOperator, EqualsOrNilOperator}
+
+const (
+	// OpenBracket is the token that denotes the beginning of a multivariate operand
+	OpenBracket string = "["
+	// OpenBracket is the token that denotes the end of a multivariate operand
+	CloseBracket string = "]"
+)
+
+// CriterionType is a type of criteria to be applied when querying
+type CriterionType string
+
+const (
+	// FieldQuery denotes that the query should be executed on the entity's fields
+	FieldQuery CriterionType = "fieldQuery"
```

Comment:
It would make sense to support fieldquery out of the box/by default for new apis - e.g. without actually supporting labels. I think that currently one would need to add labels support to the resource which would introduce field querying, too - which is a bit strange

Reply:
Not necessarily. If you don't implement createLabels and updateLabels, you can add 1 line of code that checks if the criteria passed to List contains LabelQuery and you can return an error. Everything else would work fine